### DISCUSSION TITLE
#14 - Why the metric system is superior 📏

### DISCUSSION THREAD

**Discussion by jas20202** [CUMGroup/AMOGUS#122] [Blog]
Well well well,
![banana](https://github.com/CUMGroup/AMOGUS/assets/99250573/06565a1b-c5cc-45f5-842a-dce1fee6d489)

Welcome to our blog post for this week. ඞ 

Last lecture we learned a bunch of metrics to analyse ones code. Our group had the honor to present the _**"Relation Among Classes"**_ metrics to our course. This metric consists of the following 4 sub-parts. Here is a short summary: 

**Depth of the inheritance tree** 🌳
- This refers to the number of levels in a class hierarchy.  
- It is measured using a class diagram. 
- Either manually or using tools to automatically analyse it.

**Number of children** 👨‍👧‍👦
- Number of immediate sub-classes that inherit from one class. 
- It is to measure the width of the inheritance tree.
- Automation is also possible.

**Method inheritance factor** 👶
- This shows the level of reuse of a method
- It is calculated by the number of inherited methods divided by the total number of methods.

**Response for a class** 👩‍🏫
- Total number of methods that can potentially be executed in response to a message received by an object class.
- This can also be measured automatically with tools.

If you are interested, you can find all of our slides [here](https://github.com/CUMGroup/AMOGUS/blob/main/documents/metrics/Metrics-Task3-RelationAmongClasses.pdf).

<hr>

For our project we decided on the following three metrics to analyse our code:
- Source code complexity (can be calculated by our IDE) 
- Coupling (can also be calculated by our IDE)
- Web application metrics (Lighthouse Tab in developer window of the browser)

But more on that on our final project report 👀

<hr>

_**FURTHERMORE**_, we also made some progress in development 👀
1. Our parallax header now has our logo and is only in our Landing page so that it doesn't take up too much space on other views.
![image](https://github.com/CUMGroup/AMOGUS/assets/99250573/09d8eb39-2164-4a82-90e7-7e45aaee5854)
We also have a footer now with all important links, such as this blog and the GitHub repo.

2. As mentioned in our last blog we now have some sweet badges for streak milestones 🧓👵(all handmade by @jas20202)
<img src="https://github.com/CUMGroup/AMOGUS/assets/99250573/3f881a3e-9c97-46da-a208-925f08dba61b" alt="badges" width=400></img>
Your current badge is based on your longest streak and will be displayed on your stats page next to your streak 🔥
![image](https://github.com/CUMGroup/AMOGUS/assets/99250573/fa15a90f-b311-4651-8864-c69f31111cc4)

3. Last but not least LEADER-BOARDS!!! We all know that all you want is to compare yourselves to other players and show that you are the best 👀🏆
You can find three leader-boards on our landing-page when you are logged in. There our top 5 players in the categories _"longest streak"_, _"current streak"_ and _"correct to wrong ratio"_ are displayed 🦖
![image](https://github.com/CUMGroup/AMOGUS/assets/99250573/970b61b9-f802-4597-a112-7706b666ae7e)

So it's time to grab your brain cells and [speedrun](https://amogus.alexmiha.de/) up that leader-board
![You](https://media.tenor.com/St6U7CZMAM4AAAAd/rageman.gif)
> ↳ Original footage of @AlexMi-Ha in _**mental insane mode**_

That's all for now folks ✍ a pretty long, but also spicy blog _shit_ post for you💅

Yours truly,
CUMGroup

**Comment by robin13901** [comment]
Hello CUMGroup!

It's great to see your progress in both code analysis and development. The metrics you presented for analyzing code are insightful, and it's good to know that you have chosen source code complexity, coupling, and web application metrics for your project analysis. We're looking forward to reading more about them in your final project report.

On the development side, it's nice to hear about the improvements you made. The parallax header with your logo adds a nice touch to the Landing page, and the footer with important links is a useful addition. The badges for streak milestones are a fun feature, and the leaderboards add a competitive element to the platform. It seems like you're providing a well-rounded experience for your users.

Keep up the great work, and I'm excited to see what else you have in store for your project. Good luck with your coding and development endeavors!

Best regards,
TuDu Team

**Reply by AlexMi-Ha** [reply]
Hi Team TuDu,
thanks for the positive feedback ^^ We always strive to give the best experience _(of course with a little bit of stress, as the name suggests)_.

Yours truly,
CUMGroup

**Comment by vans1646** [comment]
Hi CUMGoup!
Kinda funny Team Capyclue also had to present the Relation among Classes metrics so I already knew all about that but also good to read it in other words again :D 

As always a really nice written blog that makes me longing for more.

Concerning the metrics you've decided on, can you tell me which IDE you're using again? We pretty much settled on all the speed metrics which are already checked by lighthouse, but maybe if the IDE can do some stuff for us we could also take some other metrics into account.

Also great to see all the progress in your project, especially those amazing handmade badges. 🤩 

Keep the good work up and we will meet again under your next blog post!

Cheers,
Capyclue

**Reply by AlexMi-Ha** [reply]
Hi Capyclue,
Great to hear from you again!
I've already read about it in your blog, that you also covered Relation among Classes ^^ It's always good to get second opinions.

We are using the Visual Studio Community 2022 IDE. There you can just right-click your solution -> Analyse -> Create Code Metrics
![image](https://github.com/CUMGroup/AMOGUS/assets/61502536/864a751d-c479-4fb9-8d9a-3e1b9d80cf12)

There you can access multiple Source Code complexity metrics:
![image](https://github.com/CUMGroup/AMOGUS/assets/61502536/eba44051-b4b6-4950-9ee5-f424c1d0ef57)
> Keep in mind, the cyclomatic complexity for example, is summed up and should be addressed class-wise or even method-wise to give a meaningful measure (you can click on the assembly to view all classes/methods)

Glad you like the badges, @jas20202 really did something special here.

Yours truly,
CUMGroup

**Comment by tonoizer** [comment]
Hey CUMGroup,
before I talk about your metric information, I would prefer to point out that I really like what you did with your design. It looks very good, you got a very clean Design over all. Do you use something like Figma to pre sketch the designs and implement them afterwards it into your website ?

Your badges seem to be inspired by GitHub a bit ;), but this, it gives your Website a nice overall touch.

About the metrics you have decided on. I am curious, what is your source code complexity which the IDE generated for you, and how does your App perform after running Lighthouse over it ?#

I maybe have something interesting for you too. If you wanted to benchmark and test with Lighthouse, you could also have a look into [Unlighthouse](https://unlighthouse.dev/) it is like Lighthouse but better haha, just have a look at it if you are curious.

Over all, I am very impressed by your results.

Best Regards
Team Vigad

**Reply by AlexMi-Ha** [reply]
Hi Vigad,
always nice to get feedback from you!
We did not use Figma for our project, as nobody in our team used it before, so @jas20202 just quickly drew them up on their tablet.

Awesome that you caught that pinch of inspiration by the Github badges ^^ We liked them, so we wanted to make them in a similar style, but I think they turned out really really well

Our IDE Visual Studio Community 2022 can create metrics like a maintenance index, cyclomatic complexity,  depth of inheritance tree, coupling and LOC. (refer to previous comment for more insights)

I will definitely look into Unlighthouse, thanks for the suggestion  ^^

Yours truly,
CUMGroup
